pkg/actuators/machine/machine_scope.go

```diff
@@ -0,0 +1,161 @@
+	}
+
+	credentialsSecretName := ""
+	if providerSpec.CredentialsSecret != nil {
+		credentialsSecretName = providerSpec.CredentialsSecret.Name
+	}
+
+	awsClient, err := awsclient.NewClient(params.client, credentialsSecretName, params.machine.Namespace, providerSpec.Placement.Region)
+	if err != nil {
+		return nil, machineapierros.InvalidMachineConfiguration("failed to create aws client: %v", err.Error())
+	}
+
+	return &machineScope{
+		Context:            params.Context,
+		apiReader:          params.apiReader,
+		awsClient:          awsClient,
+		client:             params.client,
+		machine:            params.machine,
+		machineToBePatched: runtimeclient.MergeFrom(params.machine.DeepCopy()),
+		providerSpec:       providerSpec,
+		providerStatus:     providerStatus,
+	}, nil
+}
+
+// Patch patches the machine spec and machine status after reconciling.
+func (s *machineScope) patchMachine() error {
+	klog.V(3).Infof("%v: patching machine", s.machine.GetName())
+
+	providerStatus, err := awsproviderv1.RawExtensionFromProviderStatus(s.providerStatus)
+	if err != nil {
+		return machineapierros.InvalidMachineConfiguration("failed to get machine provider status: %v", err.Error())
+	}
+	s.machine.Status.ProviderStatus = providerStatus
+
+	statusCopy := *s.machine.Status.DeepCopy()
+
+	// patch machine
+	if err := s.client.Patch(context.Background(), s.machine, s.machineToBePatched); err != nil {
+		klog.Errorf("Failed to patch machine %q: %v", s.machine.GetName(), err)
+		return err
+	}
+
+	s.machine.Status = statusCopy
+
+	// patch status
+	if err := s.client.Status().Patch(context.Background(), s.machine, s.machineToBePatched); err != nil {
+		klog.Errorf("Failed to patch machine status %q: %v", s.machine.GetName(), err)
+		return err
+	}
+
+	return nil
+}
+
+// getUserData fetches the user-data from the secret referenced in the Machine's
+// provider spec, if one is set.
+func (s *machineScope) getUserData() ([]byte, error) {
+	if s.providerSpec == nil || s.providerSpec.UserDataSecret == nil {
+		return nil, nil
+	}
+
+	userDataSecret := &corev1.Secret{}
+
+	objKey := runtimeclient.ObjectKey{
+		Namespace: s.machine.Namespace,
+		Name:      s.providerSpec.UserDataSecret.Name,
+	}
+
+	if err := s.client.Get(s.Context, objKey, userDataSecret); err != nil {
+		return nil, err
+	}
+
+	userData, exists := userDataSecret.Data[userDataSecretKey]
+	if !exists {
+		return nil, fmt.Errorf("secret %s missing %s key", objKey, userDataSecretKey)
+	}
+
+	return userData, nil
+}
+
+func (s *machineScope) setProviderStatus(instance *ec2.Instance, condition awsproviderv1.AWSMachineProviderCondition) error {
+	klog.Infof("%s: Updating status", s.machine.Name)
+
+	// Save this, we need to check if it changed later.
+	networkAddresses := []corev1.NodeAddress{}
+
+	// Instance may have existed but been deleted outside our control, clear it's status if so:
+	if instance == nil {
+		s.providerStatus.InstanceID = nil
+		s.providerStatus.InstanceState = nil
```

Comment:
In master branch code this would only be reached for edge case where an instance is removed out of band and goes away during the update call.
Even for such scenario we should not clean status so this seems a bug to me. I'd suggest we keep it though to contain the scope of the PR.

Reply:
@alexander-demichev if you both agree on the above can you add a TODO here?

Reply:
I can remove it when this PR gets merged